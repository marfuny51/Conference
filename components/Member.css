.Table {
    text-align: center;
}

.TableMember {
    border-collapse: collapse;
    width: 70vw;
    height: 10vh;
    margin: 6vh auto 2vh auto;
}

.TableMemberAdd {
    border-collapse: collapse;
    width: 70vw;
    height: 10vh;
    margin: 6vh auto 2vh auto;
    opacity:0.3;
}

.TableHead {
    color: #00D970;
}

.TableHead th {
    font-family: 'Muli', sans-serif;
    font-size: 1.2vw;
    background-color: #127a16;
    color: white;
    letter-spacing: 0.1vw;
    padding: 1vw;
    border: 0vw;
}

.TableBody tr {
    height: 6vh;
    background-color: #d3f1d4;
    border-bottom: 0.15vw solid #127a16;
}

.TableBody tr:hover {
    background-color: #f5f5f5;
}

.TableBody tr td {
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    border: 0vw;
    padding: 0.5vw;
}

.Read {
    font-family: 'Muli', sans-serif;
    font-size: 1vw;
    height: 4.2vh;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
}

.Read:hover {
    font-weight: bold;
    transform: scale(1.03);
    border:0.2vw solid #456835;
    cursor: pointer;
}

.Read2 {
    font-family: 'Muli', sans-serif;
    font-size: 1vw;
    height: 4.2vh;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
    margin-right: 1.2vw;
}

.Read2:hover {
    font-weight: bold;
    transform: scale(1.03);
    border:0.2vw solid #456835;
    cursor: pointer;
}

.Add {
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    font-weight: bold;
    height: 6.5vh;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
}

.Add:hover {
    transform: scale(1.03);
    border:0.2vwsolid #456835;
    cursor: pointer;
}

.NavLink {
    color: #172c11;
    font-weight: bold;
}

.LoadData {
    font-size: 1.2vw;
    text-align: center;
    color: white;
}

.AddMember {
    width: 48vw;
    height: 32vh;
    position: absolute;
    top: 35%;
    left: 26%;
    background-color: #d3f1d4;
    border: 0.2vw solid #77b45b;
    border-radius: 0.5vw;
}

.Title {
    font-family: 'Muli', sans-serif;
    font-weight: bold;
    letter-spacing:0.1vw;
    padding-top: 1vw;
}

.Input {
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    text-align: left;
    padding: 3.8vh 1vw 3vh 4vw;
}

.Input input{
    width: 9vw;
    height: 2.5vh;
}

.Valide {
    color: red;
}
