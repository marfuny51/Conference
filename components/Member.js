import React from 'react';
import PropTypes from 'prop-types';
import {voteEvents} from './events';

import './Member.css';

class Member extends React.PureComponent {

  static propTypes = {
    member:PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      phone: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      job: PropTypes.string.isRequired,
    }),
  };

  deleteMember = () => {
    let question = confirm('Do you sure? You are going to delete this information.');
    if(question) {
      voteEvents.emit('EDeleteMember', this.props.member.id);
    }
  }

  editMember = () => {
    voteEvents.emit('EEditMember', this.props.member.id);
  }

  render() {
    
    return (
      <tr key={this.props.member.id}>
        <td>{this.props.member.name}</td>
        <td>{this.props.member.phone}</td>
        <td>{this.props.member.email}</td>
        <td>{this.props.member.job}</td>
        <td><input type="button" value="Edit" className='Read' disabled = {(this.props.mode===8)?true:false} onClick={this.editMember}/></td>
        <td><input type="button" value="Delete" className='Read' disabled = {(this.props.mode===8)?true:false} onClick={this.deleteMember}/></td>
      </tr>
    );

  }

}

export default Member;
