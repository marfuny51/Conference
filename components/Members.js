import React from 'react';
import PropTypes from 'prop-types';
import {voteEvents} from './events';
import isoFetch from 'isomorphic-fetch';
import Member from './Member';
import AddMember from './addMember';
import EditMember from './editMember';

import './Topics.css';

class Members extends React.PureComponent {

  /*constructor(props) {
    super(props);
    this.mounted = false;
  }*/

  state = {
    dataReady: false,
    members: [],
    editCode: null,
    mode: null, //6 - view, 7 - edit, 8- add
  }

  membersArray;

  componentDidMount = () => {
    this.loadData();
    //this.mounted = true;
    voteEvents.addListener('ESaveMember',this.memberSave);
    voteEvents.addListener('ECancelMember',this.memberCancel);
    voteEvents.addListener('EDeleteMember',this.delete);
    voteEvents.addListener('EEditMember',this.memberEdit);
    voteEvents.addListener('EEditSaveMember',this.memberSave);     
  };

  componentWillUnmount = () => {
    //this.mounted = false;
    voteEvents.removeListener('ESaveMember',this.memberSave);
    voteEvents.removeListener('ECancelMember',this.memberCancel);
    voteEvents.removeListener('EDeleteMember',this.delete);
    voteEvents.removeListener('EEditMember',this.memberEdit);
    voteEvents.removeListener('EEditSaveMember',this.memberSave);
  };

  loadData = () => {
    let ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    let sp = new URLSearchParams();
    sp.append('f', 'READ');
    sp.append('n', 'PROBA_PERA7');

    isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp,
    })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result === "") {
              this.speakersArray = {};
              this.speakersArray.speakers = [];
              this.speakersArray.topics = [];
              this.speakersArray.members = [];
            }
            else this.speakersArray = JSON.parse(data.result);
            console.log(this.speakersArray.members);
            
              this.setState({ dataReady:true, members:this.speakersArray.members});
            
        })
        .catch( error => {
            this.fetchError(error.message);
        });

  }; 

  fetchError = (errorMessage) => {
    console.error(showStr);
  };

  memberSave = (id, name, phone, email, job) => {
    var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    var updatePassword=Math.random();
    let sp1 = new URLSearchParams();
    sp1.append('f', 'LOCKGET');
    sp1.append('n', 'PROBA_PERA7');
    sp1.append('p', updatePassword);
    
    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp1,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( data => {
          if (data.result === "") {
            this.speakersArray = {};
            this.speakersArray.members = [];
          }
          if (data.result.members === undefined||data.result.members==='') {
            this.speakersArray.members =[];
          }
          else this.speakersArray.members = JSON.parse(data.result.members);
      })
      .catch( error => {
          console.log(error.message);
      });
    
      if (this.state.mode === 7) {
        console.log(name, phone, email, job);
        this.speakersArray.members.forEach( member => {
          if (member.id === id) {
            member.name = name;
            member.phone = phone;
            member.email = email;
            member.job = job;
          }
        })
      }

      else if (this.state.mode === 8) {
        this.speakersArray.members.push({id: id, name: name, phone: phone, email: email, job: job});
        console.log(this.speakersArray);
      }

    let sp2 = new URLSearchParams();
    sp2.append('f', 'UPDATE');
    sp2.append('n', 'PROBA_PERA7');
    sp2.append('p', updatePassword);
    sp2.append('v', JSON.stringify(this.speakersArray));

    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp2,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( (data) => {
        console.log(data);
        this.loadData();
        this.setState({mode:6});
      })
      .catch( error => {
          this.fetchError(error.message);
      });
    }

    delete = (id) => {
      var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
      var updatePassword=Math.random();
      let sp3 = new URLSearchParams();
      sp3.append('f', 'LOCKGET');
      sp3.append('n', 'PROBA_PERA7');
      sp3.append('p', updatePassword);
      
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp3,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result === "") {
              this.speakersArray = {};
              this.speakersArray.members = [];
            }
            if (data.result.members === undefined||data.result.members==='') {
              this.speakersArray.members =[];
            }
            else this.speakersArray = JSON.parse(data.result);
        })
        .catch( error => {
            console.log(error.message);
        });
      
      this.speakersArray.members = this.speakersArray.members.filter(member => member.id !== id);
      console.log(this.speakersArray);
  
      let sp2 = new URLSearchParams();
      sp2.append('f', 'UPDATE');
      sp2.append('n', 'PROBA_PERA7');
      sp2.append('p', updatePassword);
      sp2.append('v', JSON.stringify(this.speakersArray));
  
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp2,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( (data) => {
          console.log(data);
          this.loadData();
        })
        .catch( error => {
            this.fetchError(error.message);
        });
      }

  memberCancel = () => {
    let members = [...this.state.members];
    this.setState({mode:6, members: members});
  }

  addMember = (EO) => {
    this.setState({mode:8})
  }

  memberEdit = (id) => {
    this.setState( {editCode:id, mode:7})
  }

  getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
  
  render() {
    if ( !this.state.dataReady )
      return <div>загрузка данных...</div>;

    let members = [...this.state.members];
    let members2 = [...this.state.members];

    var membersCode=members.map( member =>
      <Member key={member.id} 
      mode={this.state.mode}
      member={member}  />
    );

    var editCode = members2.find( member => member.id === this.state.editCode); 

    return (
      <div>
        <div>Members</div>
        <table>
        <thead>
              <tr>
                  <th>Name</th>
                  <th>Phone</th>
                  <th>Email</th>
                  <th>Job</th>
                  <th>Edit</th>
                  <th>Delete</th>
              </tr>
          </thead>
          <tbody>
          {membersCode}
          </tbody>
        </table>
        <input type="button" value="Add a new member" onClick = {this.addMember} disabled = {(this.state.mode===8)?true:false}/>
        {
        (this.state.mode===8)&&
          <AddMember key={this.getRandomInt(1, 10000)}
          mode={this.state.mode}
          id={this.getRandomInt(1, 10000)}
          name=''
          phone=''
          email=''
          job=''
        />
        }
        {
        (this.state.mode===7)&&
          <EditMember key={editCode.id}
          mode={this.state.mode}
          id={editCode.id}
          name= {editCode.name}
          phone={editCode.phone}
          email={editCode.email}
          job={editCode.job}/>
        }
      
      </div>
    )
    ;

  }

}

export default Members;
