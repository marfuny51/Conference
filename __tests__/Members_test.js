import React from 'react'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json'
import Member from '../components/Member'
import Members from '../components/Members'

Enzyme.configure({ adapter: new Adapter() });

const component = shallow(<Members />);
const members = [{id:4449,name:"Aleksandr Pugach",phone:"+375297598465",email:"sasha@gmail.com",job:"QA engineer"},{id:6522,name:"Ekaterina Ivanova",phone:"+375442569874",email:"katya@bk.ru",job:"HR specialist"}];
describe('<Members /> component', () => {
    it('should render', () => {
        component.setState({ dataReady: true, members: members});
        members.map( member =>
            shallow(<Member key={member.id} 
                member={member}/>)
          );
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();

        component.find('.Add').simulate("click");
    });

    
});

