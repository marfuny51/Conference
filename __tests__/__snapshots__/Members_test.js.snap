// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Members /> component should render 1`] = `
<Members>
  <div
    className="Table"
  >
    <table
      className="TableMember"
    >
      <thead
        className="TableHead"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Email
          </th>
          <th>
            Job
          </th>
          <th>
            Edit
          </th>
          <th>
            Delete
          </th>
        </tr>
      </thead>
      <tbody
        className="TableBody"
      >
        <Member
          key="4449"
          member={
            Object {
              "email": "sasha@gmail.com",
              "id": 4449,
              "job": "QA engineer",
              "name": "Aleksandr Pugach",
              "phone": "+375297598465",
            }
          }
          mode={null}
        >
          <tr
            key="4449"
          >
            <td>
              Aleksandr Pugach
            </td>
            <td>
              +375297598465
            </td>
            <td>
              sasha@gmail.com
            </td>
            <td>
              QA engineer
            </td>
            <td>
              <input
                className="Read"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="Edit"
              />
            </td>
            <td>
              <input
                className="Read"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="Delete"
              />
            </td>
          </tr>
        </Member>
        <Member
          key="6522"
          member={
            Object {
              "email": "katya@bk.ru",
              "id": 6522,
              "job": "HR specialist",
              "name": "Ekaterina Ivanova",
              "phone": "+375442569874",
            }
          }
          mode={null}
        >
          <tr
            key="6522"
          >
            <td>
              Ekaterina Ivanova
            </td>
            <td>
              +375442569874
            </td>
            <td>
              katya@bk.ru
            </td>
            <td>
              HR specialist
            </td>
            <td>
              <input
                className="Read"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="Edit"
              />
            </td>
            <td>
              <input
                className="Read"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="Delete"
              />
            </td>
          </tr>
        </Member>
      </tbody>
    </table>
    <span
      className="SelectSpan"
    >
      Перейти на страницу: 
    </span>
    <select
      className="Select"
      name="Pages"
      onChange={[Function]}
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="2"
      >
        2
      </option>
      <option
        value="3"
      >
        3
      </option>
    </select>
    <br />
    <br />
    <input
      className="Add"
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Add a new member"
    />
  </div>
</Members>
`;
